src/XmlScope.ts

```diff
@@ -153,6 +153,11 @@ export class XmlScope extends Scope {
         });
     }
 
+    public getCodeActions(file: BscFile, range: Range, codeActions: CodeAction[]) {
+        const relevantDiagnostics = this.diagnostics.filter(x => x.range?.start.line === range.start.line);
+        this.program.plugins.emit('onScopeGetCodeActions', this, file, range, relevantDiagnostics, codeActions);
```

Comment:
Should we avoid emitting the scope event if there is no diagnostic match?

Reply:
I think we should emit the event no matter what. If we are sticking with "there can be code actions that are not attached to diagnostics", then skipping scopes with no diagnostics would be detrimental to plugins that want to follow that pattern. 

We are already only emitting the event for scopes who have the associated file, so I think it should be fine to emit the event for all scopes for that file. 

For example, assume there's some MVVM plugin that does the following:
```
    <Button on:click="vm.newValue = 1 + 2 + 3" />
```

There could be a code action `Refactor to view model method` for within the `on:click`, which transforms to something like this:
```
<button on:click="vm.clickButton()"
<script>
    vm.clickButton = function()
         vm.newValue = 1 + 2 + 3
    end function
</script>
```

Now, an argument can still be made that there should be a "hint" diagnostic there to begin with, but that's what we need to decide. **Should brighterscript require codeActions to be tied to a diagnostic?**

Reply:
Some code actions surely shouldn't be a "compilation cost" and be late evaluated if possible.

BTW the compiler should have a mode so that for pure compilation there would be no work done to feed the metadata for codeactions.

Reply:
Again, are we optmizing early here? What is the cost of these guards (coneptionally, and in terms of implementation)? It can't be overstated: Apps can never be bigger than 3mb. What else are we using the compiler for? Is it worth all the extra logic for a few 50ms on a non-ide bsc transpile? 

I think we're better off solving the speed and stability issues, and improving the api than imposing more constraints and adding more logic.

Reply:
> Some code actions surely shouldn't be a "compilation cost" and be late evaluated if possible.

This would definitely be a justification for NOT mandating a diagnostic for every code action (and therefore we must emit the event for all scopes for the file).  In that way, we don't need a special compilation mode then either.

if a plug-in wants to generate tons of code actions, they can just do it on demand when the ide asks for that specific range. That costs no extra compile time which IS something we should be guarding. 

So I'm inclined to agree with George's statement that all of this discussion around it to prevent 
few eventss from firing might be a bit of premature optimization. 

I will do a little testing to see what the performance impact of many scopes for a file would.hsve on code actions, and we can make a more educated decision based on that.

Reply:
I think I said that I'm fine with dispatching the events. Keep cool 😅

Reply:
All cool! 😁 Just thinking out loud. I appreciate all the feedback. You know me, I just want to find the best solution.